Pre release

andrewinc opened the pull request:
samples for distributed counters

andrewinc committed:
Pre release

IlyaFaer commented on firestore/src/firestore_snippets/SolutionCounters.php:
```diff
@@ -0,0 +1,142 @@
+class SolutionCounters {
+    
+    const SHARD_COLLECT_NAME='SHARDS';
+    const SHARDS_COUNT=10;
```
Это не очень хорошо. У тебя, получается, на сайте отобразится код, который после тега в строке 78, а SHARDS_COUNT и SHARD_COLLECT_NAME инстанцированы до него. Пользователь их не увидит. Имя коллекции лучше сделать просто "shard" в коде, а кол-во брать из `Counter.numShards` - как в других сэмплах

andrewinc replied:
Принято. перепишу без констант.

IlyaFaer commented on firestore/src/firestore_snippets/SolutionCounters.php:
````diff
@@ -0,0 +1,142 @@
+     * Example:
+     * ``` 
+     * SolutionCounters::incrementCounter($db->collection('MyCollection')->document('CounterDoc'), $counter->getNumShards());
+     * ```
````
Example не нужны, сэмплы сами по себе являются code example

andrewinc replied:
Да.. согласен. Проглядел. Прошу "понять и простить" &copy;

IlyaFaer commented on firestore/test/SolutionCountersTest.php:
```diff
@@ -0,0 +1,126 @@
+    public function testCounter()
+    {
+        $counter=new Counter(1000);
+        $this->assertEquals(1000, $counter->getNumShards());
```
1000 - это очень много для теста. 2-3-х шардов будет достаточно

andrewinc replied:
Так там же в реальности только значение проверяется, а не запрос с созданием 1000 документов делается. Смысл прост:
1. Создали объект с единственным полем.
2. Посмотрели правильно ли записано это поле.

В таком случае что 3, что 3000, что миллион. Вызывается один assert - ему по-барабану. Никаких действий с google-api не происходит.

IlyaFaer commented on firestore/src/firestore_snippets/SolutionCounters.php:
```diff
@@ -0,0 +1,142 @@
+        }
+    }
+    
+    // [END fs_create_counter]
```
Если не форматер проставляет пустые строки до и после тегов START-END, лучше убрать их. А то на странице с сэмплами появятся большие отступы

andrewinc replied:
Я просто подглядел как ты расставлял эти теги. Старался выдержать тот же стиль.

IlyaFaer commented on firestore/src/firestore_snippets/SolutionCounters.php:
```diff
@@ -0,0 +1,142 @@
+     * @param DocumentReference $ref Firestore document
+     * @param int $numShards The number of counter fragments. (default 10)
+     */
+    public static function incrementCounter(DocumentReference $ref, int $numShards=self::SHARDS_COUNT)
```
Два аргумента? Должен быть один - документ, а `numShards` стоит брать из самого объекта `Count`

andrewinc replied:
Тогда потребуется:
1. Либо передавать `Count `вместо `numShards `вторым аргументом. И тогда не получается избавиться от второго аргумента.
2. Либо отказаться от статических методов и создавать экземпляр объекта в котором хранить созданный при инициализации `Count`.

С одной стороны просится второй вариант, если нужно отказаться от второго аргумента, но с другой - какой-же это счётчик, скажем на вэб-странице если перед инкрементом нужно вызвать метод обнуляющий все ранее накопленные сведения. Я, признаться, удручён размышляя о практическом его применении.
Но если такова политика... Завтра же "запилю" вариант без статических методов.

andrewinc committed:
last static

andrewinc committed:
metods moved to class Counter

IlyaFaer commented on firestore/src/firestore_snippets/Counter.php:
```diff
@@ -0,0 +1,125 @@
+        return $this->count;
+    }
+}
+
```
Ну, вот здесь ещё лишняя пустая строка

IlyaFaer reviewed (approved):
Одно замечание, и всё, можно отправлять

andrewinc committed:
removed empty strings

IlyaFaer commented:
@andrewinc, ну, всё, давай пушить в Гугл

andrewinc committed:
firestore distributed counters
- ADD 3 file (command) and test method testDistributedCounter for them
- remove folder firestore_snippets with file
- remove test CounterTest.php


 an3 file and
